File: serial_comm/python/req_handler.py
import serial
import requests

ser = serial.Serial("/dev/ttyACM0", 9600, timeout=1) # Remember to change port number.

# API endpoint URL
api_url = "http://localhost:1000/health-track/record"

try:
    while True:
        data = ser.readline().decode().strip()
        if data:
            print(data);
            if "Patient ID:" in data and "Heart Rate:" in data and "Body Temperature:" in data:
                # Split the data string into components
                components = data.split(" && ")
                # Extract patient ID, heart rate and body temperature
                patient_id = components[0].split(": ")[1]
                heart_rate = float(components[1].split(": ")[1])
                body_temperature = float(components[2].split(": ")[1])

                if 40 <= heart_rate <= 160:
                    request_body = {"patient_id": patient_id, "heart_rate": int(float(heart_rate)), "body_temperature": int(float(body_temperature))}
                    response = requests.post(api_url, json=request_body)
                    if response.status_code == 201:
                        print(response.json())
                    else:
                        print(response.json())
                        print("Failed to upload data")
except KeyboardInterrupt:
    print("Serial communication interrupted.")
except requests.exceptions.RequestException as e:
    print(f"Error: {e}")
finally:
    ser.close()

File: serial_comm/nodejs/req_handler.js
import { SerialPort } from 'serialport';
const Readline = SerialPort.parsers.Readline;
import { request } from 'request';
import process from 'process';


const port = new SerialPort('/dev/ttyACM0', {
  baudRate: 9600,
});

const parser = port.pipe(new Readline({ delimiter: '\r\n' }));

const apiURL = 'http://localhost:1000/health-track/record';

parser.on('data', (data) => {
  data = data.trim();
  if (data && parseFloat(data) <= 120 && parseFloat(data) >= 60) {
    console.log(`Received data from Arduino: ${data}`);
    const request_body = {
      patient_id: '4567893021869403',
      heart_rate: parseInt(parseFloat(data)),
      body_temperature: 42,
    };

    request.post(
      {
        url: apiURL,
        json: request_body,
      },
      (error, response) => {
        if (!error && response.statusCode === 201) {
          console.log('Data uploaded successfully!');
        } else {
          console.log('Failed to upload data');
        }
      }
    );
  }
});

process.on('SIGINT', () => {
  console.log('Serial communication interrupted.');
  port.close();
});